### ISSUE TITLE
Replace ABSPATH in lightbox-photoswipe.php

### ISSUE THREAD

**Issue by samirillion** [arnowelzel/lightbox-photoswipe#33]
ABSPATH caused some problems for me on Flywheel servers because Flywheel uses symlinks.

Per [the docs](https://developer.wordpress.org/plugins/plugin-basics/determining-plugin-and-content-directories/ ) it seems that plugins should avoid constants like ABSPATH in general.

Great plugin!

Sam

**Comment by arnowelzel** [comment]
Thanks for pointing this out. My problem is, that I need to resolve the full path of an image file based on its URL. Using `wp_upload_dir()` won't work, since this returns the full path to the uploads folder - but I only need the base folder of WordPress itself.

Example: an image has the path `/wp-content/uploads/2020/05/myimage.jpg`. To be able to read this file `ABSPATH` is used to get the full path to that image:

`ABSPATH . '/wp-content/uploads/2020/05/myimage.jpg'`

One option might be to use `get_home_path()` - but this function is not available without including some other files first... based on the absolute path of the WordPress installation m(.

So I don't know how to change this at the moment.

**Comment by arnowelzel** [comment]
I realized that even `wp_upload_dir()` may use `ABSPATH`, so even if I'd manage to use that function, it may still not work: https://developer.wordpress.org/reference/functions/wp_upload_dir/

What **exaclty** is the problem with Flywheel hosted websites?

**Comment by samirillion** [comment]
> What exaclty is the problem with Flywheel hosted websites?

**per my back-and-forth with Flywheel:**
> `wordpress` is a symlinked folder in the site root where we keep the WP Core files. Sometimes you see this pop into URIs when it's assumed that wp-content is in the same directory as other WP Core files...I would definitely use something more specific when determining paths to mitigate the point of failure as others move to dynamic infrastructures "Cloud Hosting".

I'm agnostic about this server configuration, since I really don't know if this is "doing it right" or not.

> I realized that even wp_upload_dir() may use ABSPATH

So the funny thing is `wp_upload_dir()` seems to have the same weaknesses, looks like `ABSPATH` is used in that function, which may be a bug, or at least a design problem. This may be a WordPress problem, since they say "don't use constants," but then they put constants in some of these functions. Based on my research, `wp_upload_dir()` has had a few different problems in the past: not working with https, not working on multisite, etc.

I think using `dirname()` may be the way to go? `plugin_dir_path()`, for example, is just a wrapper around `dirname()`:
```
function plugin_dir_path( $file ) {
    return trailingslashit( dirname( $file ) );
}
```

 I'm not proud of this, but to get photoswipe working, I copied the code into a theme and am using `dirname(dirname(dirname(get_stylesheet_directory())));`, which gives me the correct path.

If you want, I can take the time to write a function to handle this case a little better.

**Comment by arnowelzel** [comment]
Thank you for the additional information. I'll see if I can set up a similar enviroment on my development server so I can reproduce the effect.

I would help a lot if you could test a new version before I release it. At the moment I am already working on a complete new backend design with a tabbed interface (without the need of reloading when switching tabs of course ;-)) and I would like to implement the fix there instead of releasing another "old" version.

![grafik](https://user-images.githubusercontent.com/6613614/81839794-508fb280-9548-11ea-85d3-0d0ed1a87d24.png)

**Comment by arnowelzel** [comment]
After playing around a bit with it - the hack using `dirname(dirname(dirname(...)))` also not suitable for all cases. This breaks if the plugin itself or the plugin folder is included using a symbolic link because in this case dirname() will also report the wrong folder.

Also the code of `get_home_path()` will not work as it also uses `ABSPATH` in certain situations.

So far using the path of `$_SERVER['SCRIPT_FILENAME']` seems to be the most promising way. But this is also just a hack and will also break if `wp-content` is not located in that path.

**Comment by arnowelzel** [comment]
Sorry for the looooong delay. I finally found a way mich might work. Can you please test, if my fix in 8054bc1c46820c233f3ae8296d02bdb761c0ca5a works for you? If yes, I will include this in the next update - thank you!
